compiler/util/parser_types.py

```diff
@@ -12,72 +12,139 @@
+        assert isinstance(line, int), f"line {line} is not int"
+        assert isinstance(column, int), f"column {column} is not int"
+        assert line >= 0, f"line {line} is negative"
+        assert column >= 0, f"column {column} is negative"
+        return PositionTuple.__new__(cls, line, column)
 
+    def __str__(self):
+        return f"{self.line}:{self.column}"
 
-def make_location(start, end, is_synthetic=False):
-    """Makes an ir_data.Location from (line, column) tuples or ir_data.Positions."""
-    if isinstance(start, tuple):
-        start = _make_position(*start)
-    if isinstance(end, tuple):
-        end = _make_position(*end)
-    if not isinstance(start, ir_data.Position):
-        raise ValueError("Bad start {!r}".format(start))
-    elif not isinstance(end, ir_data.Position):
-        raise ValueError("Bad end {!r}".format(end))
-    elif start.line > end.line or (
-        start.line == end.line and start.column > end.column
+    @staticmethod
+    def from_str(value):
+        try:
+            l, c = value.split(":")
+            return SourcePosition(line=int(l.strip()), column=int(c.strip()))
+        except Exception as e:
+            raise ValueError(f"{repr(value)} is not a valid SourcePosition.")
+
+    def __bool__(self):
+        return bool(self.line or self.column)
+
+
+LocationTuple = collections.namedtuple(
+    "LocationTuple",
+    ["start", "end", "is_disjoint_from_parent", "is_synthetic"],
+    defaults=[SourcePosition(), SourcePosition(), False, False],
+)
+
+
+class SourceLocation(LocationTuple):
+    """A half-open start:end range within a source file."""
+
+    def __new__(
+        cls,
+        /,
+        start=SourcePosition(),
+        end=SourcePosition(),
+        *,
+        is_disjoint_from_parent=False,
+        is_synthetic=False,
     ):
-        raise ValueError(
-            "Start {} is after end {}".format(
-                format_position(start), format_position(end)
-            )
+        if not isinstance(start, SourcePosition):
+            start = SourcePosition(*start)
+        if not isinstance(end, SourcePosition):
+            end = SourcePosition(*end)
+        assert start <= end, f"start {start} is after end {end}"
+        assert isinstance(
+            is_disjoint_from_parent, bool
+        ), f"is_disjoint_from_parent {is_disjoint_from_parent} is not bool"
+        assert isinstance(
+            is_synthetic, bool
+        ), f"is_synthetic {is_synthetic} is not bool"
+        return LocationTuple.__new__(
+            cls, start, end, is_disjoint_from_parent, is_synthetic
         )
-    return ir_data.Location(start=start, end=end, is_synthetic=is_synthetic)
 
+    def __str__(self):
+        suffix = ""
+        if self.is_disjoint_from_parent:
+            suffix += "^"
+        if self.is_synthetic:
+            suffix += "*"
+        return (
+            f"{self.start or SourcePosition()}-{self.end or SourcePosition()}{suffix}"
```

Comment:
The `or SourcePosition()` shouldn't be necessary here right?

Reply:
Done.

Left over from a previous iteration, when I allowed `start` and `end` to be `None`.